### ISSUE TITLE
Fails to build on linux arm

### ISSUE THREAD

**Issue by elmart-devo** [VectorCamp/vectorscan#146]
Hi, I'm getting linking problems when trying to build on a linux arm machine.

I notice in the forefront summary that there's some issue determining flags:
```
-- The C compiler identification is GNU 9.4.0
-- The CXX compiler identification is GNU 9.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Build type RELWITHDEBINFO
-- using release build
-- Boost version: 1.71.0
-- Build date: 2023-04-18
-- Building shared libraries
-- gcc version 9.4.0
CMake Warning at CMakeLists.txt:184 (message):
-- ARCH_C_FLAGS   : 
  Something went wrong determining gcc tune:
-- ARCH_CXX_FLAGS : 
  -mtune=armv8.2-a+crypto+fp16+rcpc+dotprod not valid, falling back to
-- g++ version 9.4.0
  -mtune=native
```

Nonetheless, it continues compiling and linking everything until linking `../bin/util/internal` (at 100% progress):
```
[ 93%] Building CXX object unit/CMakeFiles/unit-internal.dir/internal/rose_mask.cpp.o
[ 93%] Building CXX object unit/CMakeFiles/unit-internal.dir/internal/rose_mask_32.cpp.o
[ 94%] Building CXX object unit/CMakeFiles/unit-internal.dir/internal/rvermicelli.cpp.o
[ 94%] Building CXX object unit/CMakeFiles/unit-internal.dir/internal/simd_utils.cpp.o
[ 94%] Building CXX object unit/CMakeFiles/unit-internal.dir/internal/supervector.cpp.o
[ 95%] Building CXX object unit/CMakeFiles/unit-internal.dir/internal/shuffle.cpp.o
[ 95%] Building CXX object unit/CMakeFiles/unit-internal.dir/internal/shufti.cpp.o
[ 95%] Building CXX object unit/CMakeFiles/unit-internal.dir/internal/state_compress.cpp.o
[ 96%] Building CXX object unit/CMakeFiles/unit-internal.dir/internal/truffle.cpp.o
[ 96%] Building CXX object unit/CMakeFiles/unit-internal.dir/internal/unaligned.cpp.o
[ 96%] Building CXX object unit/CMakeFiles/unit-internal.dir/internal/unicode_set.cpp.o
[ 97%] Building CXX object unit/CMakeFiles/unit-internal.dir/internal/uniform_ops.cpp.o
[ 97%] Building CXX object unit/CMakeFiles/unit-internal.dir/internal/utf8_validate.cpp.o
[ 97%] Building CXX object unit/CMakeFiles/unit-internal.dir/internal/util_string.cpp.o
[ 98%] Building CXX object unit/CMakeFiles/unit-internal.dir/internal/vermicelli.cpp.o
[ 98%] Building CXX object unit/CMakeFiles/unit-internal.dir/internal/main.cpp.o
[100%] Linking CXX executable ../bin/unit-internal
```
At that point, it starts emitting errors like these:
```
/usr/bin/ld: CMakeFiles/unit-internal.dir/internal/database.cpp.o: in function `CRC_alignments_Test::TestBody()':
/var/lib/jenkins/workspace/hs-native_feature_vectorscan/.gradle/vcs-1/vectorscan_95duba3co85x2wvqcmpl4rcyt/vectorscan/unit/internal/database.cpp:69: undefined reference to `Crc32c_ComputeBuf'
/usr/bin/ld: CMakeFiles/unit-internal.dir/internal/database.cpp.o: in function `DB_flagsToPlatform_Test::TestBody()':
/var/lib/jenkins/workspace/hs-native_feature_vectorscan/.gradle/vcs-1/vectorscan_95duba3co85x2wvqcmpl4rcyt/vectorscan/unit/internal/database.cpp:59: undefined reference to `ue2::target_t::target_t(hs_platform_info const&)'
/usr/bin/ld: /var/lib/jenkins/workspace/hs-native_feature_vectorscan/.gradle/vcs-1/vectorscan_95duba3co85x2wvqcmpl4rcyt/vectorscan/unit/internal/database.cpp:59: undefined reference to `ue2::target_to_platform(ue2::target_t const&)'
/usr/bin/ld: CMakeFiles/unit-internal.dir/internal/fdr_loadval.cpp.o: in function `ue2::bytecode_ptr<unsigned char>::bytecode_ptr(unsigned long, unsigned long)':
/var/lib/jenkins/workspace/hs-native_feature_vectorscan/.gradle/vcs-1/vectorscan_95duba3co85x2wvqcmpl4rcyt/vectorscan/src/util/bytecode_ptr.h:63: undefined reference to `ue2::aligned_malloc_internal(unsigned long, unsigned long)'
/usr/bin/ld: CMakeFiles/unit-internal.dir/internal/fdr_loadval.cpp.o: in function `ue2::bytecode_ptr<unsigned char>::deleter<unsigned char>::operator()(unsigned char*) const':
/var/lib/jenkins/workspace/hs-native_feature_vectorscan/.gradle/vcs-1/vectorscan_95duba3co85x2wvqcmpl4rcyt/vectorscan/src/util/bytecode_ptr.h:130: undefined reference to `ue2::aligned_free_internal(void*)'
/usr/bin/ld: /var/lib/jenkins/workspace/hs-native_feature_vectorscan/.gradle/vcs-1/vectorscan_95duba3co85x2wvqcmpl4rcyt/vectorscan/src/util/bytecode_ptr.h:130: undefined reference to `ue2::aligned_free_internal(void*)'
/usr/bin/ld: /var/lib/jenkins/workspace/hs-native_feature_vectorscan/.gradle/vcs-1/vectorscan_95duba3co85x2wvqcmpl4rcyt/vectorscan/src/util/bytecode_ptr.h:130: undefined reference to `ue2::aligned_free_internal(void*)'
/usr/bin/ld: /var/lib/jenkins/workspace/hs-native_feature_vectorscan/.gradle/vcs-1/vectorscan_95duba3co85x2wvqcmpl4rcyt/vectorscan/src/util/bytecode_ptr.h:130: undefined reference to `ue2::aligned_free_internal(void*)'
/usr/bin/ld: CMakeFiles/unit-internal.dir/internal/fdr_loadval.cpp.o: in function `ue2::bytecode_ptr<unsigned char>::bytecode_ptr(unsigned long, unsigned long)':
/var/lib/jenkins/workspace/hs-native_feature_vectorscan/.gradle/vcs-1/vectorscan_95duba3co85x2wvqcmpl4rcyt/vectorscan/src/util/bytecode_ptr.h:63: undefined reference to `ue2::aligned_malloc_internal(unsigned long, unsigned long)'
```

Any help, please???

**Comment by markos** [comment]
hi, could you mention the hardware & distribution/version used so that I can try to reproduce?

**Comment by elmart-devo** [comment]
Sure. The arm machine is an AWS instance (m6g.2xlarge), with a graviton2 processor. Some more info here: https://developer.arm.com/documentation/100616/0400/. That is because failing build is running in a Jenkins CI pipeline, which builds both on intel and arm. My local machine is intel, so I can't test locally.

System is Ubuntu 20.04. In fact, failing build runs within a docker container created a dockerfile similar to this (I've edited it to remove noise):
```
FROM ubuntu:20.04
ARG USER_ID=9001
ARG GROUP_ID=9001
ARG WORK_DIR=/hs
ENV LANG=C.UTF-8
# hyperscan deps needed to compile hyperscan from source
RUN apt update \
    && DEBIAN_FRONTEND=noninteractive \
       apt install -y pkg-config g++ cmake libboost-all-dev ragel

# give user sudo privileges to default user so that
# customized child images can install things on their own
RUN apt install -y sudo \
    && echo '%sudo ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers

# create user/group
RUN groupadd --gid $GROUP_ID hs \
    && useradd \
         --create-home --no-log-init \
         --uid $USER_ID --gid $GROUP_ID --groups sudo hs

USER hs
WORKDIR $WORK_DIR
```
There you can also see installed software/etc.

**Comment by markos** [comment]
one more question, the cmake invocation please

**Comment by elmart-devo** [comment]
I'm using vectorscan 5.4.8. I hadn't realized there was a newer one. Will try that now and see what happens.

Regarding the cmake invocation, I will have to investigate a little bit, as I am not invoking cmake myself, but letting gradle do it through its native support. Will return with an answer asap.

**Comment by elmart-devo** [comment]
Ok, I've tried with vectorscan 5.4.9 and result is the same.

Let $BASE be the directory where vectorscan sources are downloaded. Then:

CMake invocation is:
```
(at working dir $BASE/build/release)
cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo -DBUILD_SHARED_LIBS=ON -DCMAKE_INSTALL_PREFIX=$BASE/build/release/install $BASE```
```
That seems to generate cmake files ok.
Then I execute make:
```
(at working dir $BASE/build/release)
make -j8
```
As I said before, I'm not really doing that directly, but through gradle instead. But I don't think the problem is in the gradle integration, as the same thing works fine for intel platform, and the error output at the end of all link problems like posted above says this:
```
/usr/bin/ld: /var/lib/jenkins/workspace/hs-native_feature_vectorscan/.gradle/vcs-1/vectorscan_3na5pbvzmno96zke757i6zhk7/vectorscan/unit/internal/util_string.cpp:410: undefined reference to `ue2::ue2_literal::erase(unsigned long, unsigned long)'
/usr/bin/ld: /var/lib/jenkins/workspace/hs-native_feature_vectorscan/.gradle/vcs-1/vectorscan_3na5pbvzmno96zke757i6zhk7/vectorscan/unit/internal/util_string.cpp:411: undefined reference to `ue2::ue2_literal::ue2_literal(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool)'
collect2: error: ld returned 1 exit status
make[2]: *** [unit/CMakeFiles/unit-internal.dir/build.make:776: bin/unit-internal] Error 1
make[1]: *** [CMakeFiles/Makefile2:493: unit/CMakeFiles/unit-internal.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
```
Which makes me think that the problem is in the build itself.
So, could you please confirm if you are able to reproduce with these indications?

**Comment by elmart-devo** [comment]
Need anything else from me? Were you able to reproduce? Thx in advance.

**Comment by markos** [comment]
Sorry, I was in the middle of something else and between travels, I'm afraid I will be able to investigate on Monday. Apologies for inconvenience.

**Comment by elmart-devo** [comment]
Oh, no, please. Whenever you can continue working on it, I'll be grateful. Thx very much.

**Comment by elmart-devo** [comment]
Hey, I don't want to sound insisting, but do you think you'll find the time to check this soon?

**Comment by markos** [comment]
Hi, sorry was sidetracked by other stuff. Yes, I was able to reproduce it just now on ubuntu 22.04. I will investigate and provide a fix -along with some other stuff soon.

**Comment by graysky2** [comment]
I seem to be hit by this bug as well using Arch ARM 64-bit.  We currently have a newer toolchain (gcc 12.1.0).

If it is helpful, [here](https://gist.github.com/graysky2/6528c04839f3569c9f5fcd8a24dc3adc) is the complete build log.

I tried building like this:
```
  cmake -B build -G Ninja \
    -DCMAKE_INSTALL_PREFIX=/usr \
    -DCMAKE_INSTALL_LIBDIR=lib \
    -DBUILD_SHARED_LIBS=ON \
    -Wno-dev
  cmake --build build
```

**Comment by graysky2** [comment]
@elmart-devo - did a previous version build for you on the same system?  If so, which one?

**Comment by graysky2** [comment]
Thanks, I also did that and 5.4.7 as well but stopped trying with that.  Hopefully, @markos will have some time to fix this and bump to 5.4.10.

**Comment by elmart-devo** [comment]
Hi @markos . Any chance to get this going?
Thx very much in advance.

**Comment by markos** [comment]
Hi, sorry for the delay, this will get in next release, it's next on my TODO.

**Comment by markos** [comment]
This has been fixed in #181, please try current develop branch until new release is out. The reason was that unit-internal was being built on Shared libs builds, however that would not work as many symbols were not exposed as public. This is why builds succeeded when both STATIC and SHARED libs were built.
